cpp/include/raft/neighbors/ivf_pq_types.hpp

```diff
@@ -139,6 +142,80 @@ constexpr static uint32_t kIndexGroupSize = 32;
 
+/**
+ * Default value filled in the `indices()` array.
+ * One may encounter it trying to access a record within a cluster that is outside of the
+ * `list_sizes()` bound (due to the record alignment `kIndexGroupSize`).
+ */
+template <typename IdxT>
+constexpr static IdxT kInvalidRecord = std::numeric_limits<IdxT>::max() - 1;
+
+/**
+ * Default value returned by `search` when the `n_probes` is too small and top-k is too large.
+ * One may encounter it if the combined size of probed clusters is smaller than the requested
+ * number of results per query.
+ */
+template <typename IdxT>
+constexpr static IdxT kOutOfBoundsRecord = std::numeric_limits<IdxT>::max();
+
+/** PQ-encoded data stored in the interleaved format:
+ *
+ *    [ ceildiv(list_size, kIndexGroupSize)
+ *    , ceildiv(pq_dim, (kIndexGroupVecLen * 8u) / pq_bits)
+ *    , kIndexGroupSize
+ *    , kIndexGroupVecLen
+ *    ].
+ */
+template <typename SizeT = uint32_t>
+using list_extents =
+  extents<SizeT, dynamic_extent, dynamic_extent, kIndexGroupSize, kIndexGroupVecLen>;
+
+/** Determine the extents of an array enough to hold a given amount of data. */
+template <typename SizeT = uint32_t>
+constexpr auto make_list_extents(SizeT n_rows, uint32_t pq_bits, uint32_t pq_dim)
+  -> list_extents<SizeT>
+{
+  // how many elems of pq_dim fit into one kIndexGroupVecLen-byte chunk
+  auto pq_chunk = (kIndexGroupVecLen * 8u) / pq_bits;
+  return make_extents<SizeT>(div_rounding_up_safe<SizeT>(n_rows, kIndexGroupSize),
+                             div_rounding_up_safe<SizeT>(pq_dim, pq_chunk),
+                             kIndexGroupSize,
+                             kIndexGroupVecLen);
+}
+
+/** The data for a single list (cluster). */
+template <typename IdxT, typename SizeT = uint32_t>
+struct list_data {
+  /** PQ-encoded interleaved data. */
+  device_mdarray<uint8_t, list_extents<SizeT>, row_major> data;
+  /** Source indices. */
+  device_mdarray<IdxT, extent_1d<SizeT>, row_major> indices;
+  /** The actual size of the content. */
+  std::atomic<SizeT> size;
+
+  /** Allocate a new list capable of holding at least `n_rows` data records and indices. */
+  list_data(raft::device_resources const& res, SizeT n_rows, uint32_t pq_bits, uint32_t pq_dim)
+    : size{n_rows}
+  {
+    auto capacity = round_up_safe<SizeT>(bound_by_power_of_two<SizeT>(size), kIndexGroupSize);
+    try {
+      data = make_device_mdarray<uint8_t>(res, make_list_extents<SizeT>(capacity, pq_bits, pq_dim));
+      indices = make_device_mdarray<IdxT>(res, make_extents<SizeT>(capacity));
+    } catch (std::bad_alloc& e) {
+      RAFT_FAIL(
+        "ivf-pq: failed to allocate a big enough index list to hold all data "
+        "(requested size: %zu records, selected capacity: %zu records). "
+        "Allocator exception: %s",
+        size_t(size),
+        size_t(capacity),
+        e.what());
+    }
+    // Fill the index buffer with a pre-defined marker for easier debugging
+    thrust::fill_n(
```

Comment:
Unfortunately, I had to revert this change, because `raft::matrix::fill` is in `.cuh` file and contains cuda code, whereas `ivf_pq_types.hpp` must not.

Reply:
I still think we can take the matrix fill operation out of this file, but I don't want to hold up these changes from getting merged.